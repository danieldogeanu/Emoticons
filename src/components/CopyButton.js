import React, { Component } from 'react';
import Clipboard from 'react-clipboard.js';
import Icon from '../elements/Icon';
import { animateCSS } from '../animate';
import '../styles/components/CopyButton.scss';

class CopyButton extends Component {
	constructor(props) {
		super(props);
		this.state = {isCopied: false};
		this.copyBtn = React.createRef();
	}

	onSuccess = () => {
		this.setState({isCopied: true});
		this.copyBtn.current.element.classList.add('copied');
		setTimeout(() => {
			this.setState({isCopied: false});
			this.copyBtn.current.element.classList.remove('copied');
		}, 1500);

		this.SelectedScreen.innerHTML = this.props.data;
		animateCSS(this.SelectedScreen, 'fadeInOut');
		this.SelectedScreen.classList.add('show');
		setTimeout(() => {
			this.SelectedScreen.classList.remove('show');
		}, 1000);
	}

	componentDidMount() {
		this.SelectedScreen = document.querySelector('.SelectedScreen');
	}

	render() {
		return (
			<Clipboard
				ref={this.copyBtn}
				onSuccess={this.onSuccess}
				className={'CopyButton ' + this.props.className}
				data-clipboard-text={this.props.data}>
				{this.props.children ? this.props.children : (
					this.state.isCopied ? (<Icon name="check" />) : 'Copy'
				)}
			</Clipboard>
		);
	}
}

export default CopyButton;
