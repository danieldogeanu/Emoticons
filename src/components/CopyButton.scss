@import '../styles/includes';

.CopyButton {
	appearance: none;
	display: block;
	width: rem(55);
	height: auto;
	margin: 0;
	padding: rem(11) 0;
	border: 0 none;
	border-radius: rem(3);
	background-color: color-scheme(item, btn);
	color: color-scheme(background);
	font-size: rem(13);
	font-weight: weight(regular);
	text-transform: uppercase;
	text-align: center;
	line-height: 1;
	cursor: pointer;
	transition: background-color 0.3s ease;
	&:hover { background-color: color-scheme(item, btn-hover); }
	&:active, &:focus { outline: none; }
	&.copied {
		padding: rem(7.15) 0;
		pointer-events: none;
		user-select: none;
	}

	.icon.check {
		width: rem(21);
		height: rem(21);
		margin: 0 0 rem(-2) 0;
		padding: 0;
	}
}