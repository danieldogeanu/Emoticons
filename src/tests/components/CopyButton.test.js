import React from 'react';
import {render, fireEvent, waitForDomChange, waitForElement} from '@testing-library/react';
import CopyButton from '../../components/CopyButton';

// Mock required to trigger onSuccess event on <Clipboard> component.
document.execCommand = jest.fn(() => true);

const testEmoticon = {
	"codes": "1F604",
	"char": "ðŸ˜„",
	"name": "grinning face with smiling eyes",
	"category": "Smileys & Emotion (face-smiling)",
	"group": "Smileys & Emotion",
	"subgroup": "face-smiling"
};
const {char} = testEmoticon;

describe('CopyButton Component', () => {

	it('renders copy button properly', () => {
		const {getByTestId} = render(<CopyButton data={char} />);
		const renderedCopyBtn = getByTestId('CopyButton');

		expect(renderedCopyBtn).toBeInTheDocument();
		expect(renderedCopyBtn).toHaveClass('CopyButton', 'undefined');
		expect(renderedCopyBtn).toHaveAttribute('data-clipboard-text', char);
		expect(renderedCopyBtn).toHaveAttribute('type', 'button');
		expect(renderedCopyBtn).toHaveTextContent('Copy');
		expect(renderedCopyBtn).toMatchSnapshot();
	});

	it('renders copy button with children properly', () => {
		const {getByTestId} = render(
			<CopyButton data={char} className="children">
				<div data-testid="child">{char}</div>
			</CopyButton>
		);
		const renderedCopyBtn = getByTestId('CopyButton');
		const renderedChild = getByTestId('child');

		expect(renderedCopyBtn).toBeInTheDocument();
		expect(renderedCopyBtn).toHaveClass('CopyButton', 'children');
		expect(renderedCopyBtn).toHaveAttribute('data-clipboard-text', char);
		expect(renderedCopyBtn).toHaveAttribute('type', 'button');
		expect(renderedCopyBtn).toContainElement(renderedChild);
		expect(renderedCopyBtn).toMatchSnapshot();
		expect(renderedChild).toHaveTextContent(char);
	});

	it('copies data to clipboard', async () => {
		const {getByTestId, container} = render(
			<div>
				<CopyButton data={char} />
				<div className="SelectedScreen"></div>
			</div>
		);
		const renderedCopyBtn = getByTestId('CopyButton');
		const renderedScreen = container.querySelector('.SelectedScreen');

		fireEvent.click(renderedCopyBtn);
		const renderedIcon = await waitForElement(
			() => container.querySelector('svg'),
			{container}
		);

		// After Click Event
		expect(renderedCopyBtn).toHaveClass('copied');
		expect(renderedCopyBtn).toContainElement(renderedIcon);
		expect(renderedScreen).toHaveClass('show');
		expect(renderedScreen).toHaveTextContent(char);

		await waitForDomChange({container: renderedCopyBtn});

		// After Timers Run Out
		expect(renderedCopyBtn).not.toHaveClass('copied');
		expect(renderedCopyBtn).not.toContainElement(renderedIcon);
		expect(renderedScreen).not.toHaveClass('show');
		expect(renderedIcon).not.toBeInTheDocument();
	});

});
