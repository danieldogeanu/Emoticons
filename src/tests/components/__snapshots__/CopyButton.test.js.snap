// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CopyButton Component renders copy button properly 1`] = `
<button
  class="CopyButton undefined"
  data-clipboard-text="ðŸ˜„"
  data-testid="CopyButton"
  title=""
  type="button"
>
  Copy
</button>
`;

exports[`CopyButton Component renders copy button with children properly 1`] = `
<button
  class="CopyButton children"
  data-clipboard-text="ðŸ˜„"
  data-testid="CopyButton"
  title=""
  type="button"
>
  <div
    data-testid="child"
  >
    ðŸ˜„
  </div>
</button>
`;
